### PR TITLE
feat: wire-up RFOX change rewards address

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

TODO:

- [x] Input
  - [x] New address bech32 validation when using manual
  - [x] Confirmed quote
  - [x] Surface current address
  - [x] Empty state when no current address - TBD after discussing with @reallybeard 
  - [x] Same address not allowed for custom addy
  - [x] Same address not allowed for portfolio addy
  - [x] Gas fee
- [x] Confirm
  - [x] Gas fee
  - [x] Implement component for current/new address (TBD in a pile with @reallybeard later in the day)
  - [x] Sign and broadcast
- [x] Status
  - [x] Handle states
  - [x] Txid link

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes https://github.com/shapeshift/web/issues/6947

> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

Test address change e2e:

- Current reward saddress should be displayed
- New rewards address can be input either manually, or selected from the account dropdown list
- Choosing the current address as a new rewards address, either using the account dropdown or a manual address errors and doesn't allow to continue
- The manual receive address correctly catches same address as current and invalid RUNE address, reflected as errors
- Gas fees are happy both at status and confirm steps
- Tx goes through and status step is happy

### Engineering

☝🏽 

### Operations

☝🏽 

## Screenshots (if applicable)

https://jam.dev/c/b09f9e74-d9df-4ad5-8765-16b684717d59

### CODE REVIEW

File: src/pages/RFOX/components/ChangeAddress/ChangeAddressConfirm.tsx
Comment (gomesalexandre) [inline src/pages/RFOX/components/ChangeAddress/ChangeAddressConfirm.tsx:1]: This component is wrong - only the input part was actually implemented in the visual bits PR https://github.com/shapeshift/web/pull/6959, the confirmed part is actually a copy and paste from `<StakeConfirm />` which isn't matching the figma at all

See runtime after cleanup of ShapeShift fee and approval fees row vs. Figma 
image
image

File: src/components/AccountDropdown/AccountDropdown.tsx [hunk 1]
```diff
@@ -55,6 +55,7 @@ export type AccountDropdownProps = {
   autoSelectHighestBalance?: boolean
   // Prevents accounts in the dropdown from being selected
   disabled?: boolean
+  // Hides a given set of AccountIds from the dropdown
```
Comment (gomesalexandre) [inline src/components/AccountDropdown/AccountDropdown.tsx:58]: rm

Comment (gomesalexandre) [reply src/components/AccountDropdown/AccountDropdown.tsx]: 3e57064a

File: src/pages/RFOX/components/AddressSelection.tsx [hunk 2]
```diff
@@ -14,31 +14,45 @@ import { useCallback, useMemo, useState } from 'react'
...
-  isNewAddress?: boolean
-}
+} & (
+  | {
+      isNewAddress: boolean
+      validateIsNewAddress: (address: string) => boolean
+    }
+  | {
+      isNewAddress?: never
+      validateIsNewAddress?: never
```
Comment (gomesalexandre) [inline src/pages/RFOX/components/AddressSelection.tsx:30]: i.e both or none, but not partial

File: src/pages/RFOX/components/ChangeAddress/ChangeAddressConfirm.tsx [hunk 3]
```diff
@@ -7,62 +7,227 @@ import {
...
 
   const handleGoBack = useCallback(() => {
     history.push(ChangeAddressRoutePaths.Input)
   }, [history])
 
-  const handleSubmit = useCallback(() => {
-    history.push(ChangeAddressRoutePaths.Status)
-  }, [history])
+  const changeAddressCard = useMemo(() => {
+    // TODO(gomes): implement current/new rune address display here
```
Comment (gomesalexandre) [inline src/pages/RFOX/components/ChangeAddress/ChangeAddressConfirm.tsx:198]: Remove me

Comment (gomesalexandre) [reply src/pages/RFOX/components/ChangeAddress/ChangeAddressConfirm.tsx]: dda82383cc

File: src/pages/RFOX/components/ChangeAddress/ChangeAddressInput.tsx [hunk 4]
```diff
@@ -1,61 +1,299 @@
...
+  } = useReadContract({
+    abi: foxStakingV1Abi,
+    address: RFOX_PROXY_CONTRACT_ADDRESS,
+    functionName: 'stakingInfo',
+    args: [stakingAssetAccountAddress ? getAddress(stakingAssetAccountAddress) : ('' as Address)], // actually defined, see enabled below
+    chainId: arbitrum.id,
+    query: {
+      enabled: Boolean(stakingAssetAccountAddress),
+      // TODO(gomes): unused destructurreg values isn't an omission, it's to ensure
+      // we change it to stakingInfo[3] vs. stakingInfo[2] currently after we deploy and consume the latest version of the contract
```
Comment (gomesalexandre) [inline src/pages/RFOX/components/ChangeAddress/ChangeAddressInput.tsx:114]: To be tackled as part of https://github.com/shapeshift/web/issues/6967

File: src/pages/RFOX/components/Stake/StakeInput.tsx [hunk 5]
```diff
@@ -532,9 +524,11 @@ export const StakeInput: React.FC<StakeInputProps & StakeRouteProps> = ({
                 !cooldownPeriod
               }
               isLoading={isGetApprovalFeesLoading || isStakeFeesLoading}
-              colorScheme={Boolean(errors.amountFieldInput) ? 'red' : 'blue'}
+              colorScheme={
+                Boolean(errors.amountFieldInput || errors.manualRuneAddress) ? 'red' : 'blue'
+              }
             >
-              {errors.amountFieldInput?.message || translate('RFOX.stake')}
+              {errors.manualRuneAddress?.message || translate('RFOX.stake')}
```
Comment (gomesalexandre) [inline src/pages/RFOX/components/Stake/StakeInput.tsx:531]: This looks wrong

Comment (gomesalexandre) [reply src/pages/RFOX/components/Stake/StakeInput.tsx]: c36cf6e211

File: src/assets/translations/en/main.json [hunk 6]
```diff
@@ -2453,12 +2454,21 @@
...
+    "changeRewardAddressSuccess": "You have successfully changed your THORCHain Reward Address",
     "unstakePending": "Requesting withdraw of %{amount} %{symbol}",
     "unstakeSuccess": "Once the %{cooldownPeriod} has elapsed, you will be able to claim your %{amount} %{symbol}",
     "currentRewardAddress": "Current Reward Address",
-    "newRewardAddress": "New Address",
+    "newRewardAddress": "New Reward Address",
     "rewardCycleExplainer": "This will update for the next reward cycle.",
     "addressUpdateActionPending": "Change Reward Address...",
-    "addressUpdateSucces": "You have successfully updated your Thorchain Reward Address"
+    "addressUpdateSucces": "You have successfully updated your Thorchain Reward Address",
```
Comment (0xApotheosis) [inline src/assets/translations/en/main.json:2466]: We can remove both of these now.

File: src/assets/translations/en/main.json [hunk 7]
```diff
@@ -2453,12 +2454,21 @@
...
+    "noPositionTitle": "No Active Position",
+    "noPositionBody": "You will be able to change your reward address here, once you have an active position.",
```
Comment (0xApotheosis) [inline src/assets/translations/en/main.json:2468]: Nitpick:

```suggestion
    "noPositionBody": "You will be able to change your reward address here once you have an active position.",
```

File: src/pages/RFOX/components/Stake/StakeConfirm.tsx [hunk 8]
```diff
@@ -439,6 +439,9 @@ export const StakeConfirm: React.FC<StakeConfirmProps & StakeRouteProps> = ({
     await handleStake() // This isn't a mistake - we invalidate as a cleanup operation before unmount to avoid current subscribers refetching with wrong args, hence making invalidation useless
     history.push(StakeRoutePaths.Status)
 
+    history.push(StakeRoutePaths.Status)
```
Comment (0xApotheosis) [inline src/pages/RFOX/components/Stake/StakeConfirm.tsx:442]: Duplicate lInes?

File: src/pages/RFOX/components/ChangeAddress/ChangeAddressStatus.tsx [hunk 9]
```diff
@@ -79,7 +107,7 @@ export const ChangeAddressStatus: React.FC<ChangeAddressRouteProps> = () => {
         </AnimatePresence>
       )}
       <CardFooter flexDir='column' gap={2}>
-        <Button size='lg' variant='ghost'>
+        <Button as={Link} href={txLink} size='lg' variant='ghost'>
```
Comment (0xApotheosis) [inline src/pages/RFOX/components/ChangeAddress/ChangeAddressStatus.tsx:110]:
```suggestion
        <Button as={Link} href={txLink} size='lg' variant='ghost' isExternal>
```

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Looking great!

✅ Manual change address works as expected
✅ Change address from list works as expected
✅ Cannot proceed with same address, either manually or from list
✅ Current address displayed and updated on change

Some minor things to tweak before we get this in:

❌ Can proceed with manual address after clearing it

https://github.com/shapeshift/web/assets/97164662/3aabc271-7b26-489d-9572-a84e0e2b1908

❓ The "You have successfully..." text feels a little misaligned/missing padding. Should also probably have a full stop.

Screenshot 2024-05-27 at 2 35 13 PM

File: src/assets/translations/en/main.json (hunk 7 above)
Comment (gomesalexandre) [reply src/assets/translations/en/main.json]: 33c3c04ac821a44378d068dc1c90e39137953598

File: src/pages/RFOX/components/ChangeAddress/ChangeAddressStatus.tsx (hunk 9 above)
Comment (gomesalexandre) [reply src/pages/RFOX/components/ChangeAddress/ChangeAddressStatus.tsx]: 436f383d1dc5e58ff1f35769078fd08fce52853d

File: src/assets/translations/en/main.json (hunk 6 above)
Comment (gomesalexandre) [reply src/assets/translations/en/main.json]: 16bc6a37fb

**Comment by gomesalexandre** [comment]
Nice catch @0xApotheosis - both fixed: 
> Some minor things to tweak before we get this in:
> 
> ❌ Can proceed with manual address after clearing it
> 
>  proceed-on-empty-address.mp4 

Fixed in 29db81d08b, we were too optimistic in not firing onChange when there is no (empty) rune address. Note this commit also fixes the same issue in staking:

https://jam.dev/c/a5d65acd-1487-4250-b1db-836d87728145

> 
> ❓ The "You have successfully..." text feels a little misaligned/missing padding. Should also probably have a full stop.

Fixed in b75ff3f3cf

@woodenfurniture looks like we already have two things we improved here and want to reflect in the status steps of all three stake, unstake and change address tabs: the `isExternal` on the `<Link />`, and the text alignment improvements.

While those are not functional bugs per se, that was definitely a good call on the follow-up components restaking as a follow-up cleanup PR! 🐐

**Review by 0xApotheosis — APPROVED** [review]
Retested, looking good ser - get 'er in.

**Pull request merged**